platform/lib/routers/search.js

```diff
@@ -0,0 +1,268 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const express = require('express');
+const fs = require('fs');
+const path = require('path');
+const config = require('@lib/config.js');
+const project = require('@lib/utils/project.js');
+const googleSearch = require('@lib/utils/googleSearch.js');
+
+// google custom search does not support a page size > 10
+const PAGE_SIZE = 10;
+// google custom search json api does not support loading more than 100 results
+const LAST_PAGE = 10;
+
+const COMPONENT_REFERENCE_DOC_PATTERN =
+    /^(?:https?:\/\/[^/]+)?(?:\/[^/]+)?\/documentation\/components\/(amp-[^/]+)/;
+
+const EXAMPLE_URL_PATH = '/documentation/examples/components/';
+const EXAMPLE_FOLDER_PATH = path.join(project.paths.DIST, '/examples/sources/components/');
+
+const REMOVE_HOST_PATTERN = /^https?:\/\/amp\.dev(?=\/)/;
+
+const PREVIEW_HOST = config.hosts.preview.base;
+const PLAYGROUND_URL = config.hosts.playground.base;
+
+// use the twitter title if available since it does not contain the site name
+const TITLE_META_TAG = 'twitter:title';
+const DESCRIPTION_META_TAG = 'twitter:description';
+
+// eslint-disable-next-line new-cap
+const examples = express.Router();
+
+const BUILD_IN_COMPONENTS = [
+  'amp-layout',
+  'amp-img',
+  'amp-pixel',
+];
+
+const IMPORTANT_INCLUDED_ELEMENTS = [
+  'amp-state',
+  'amp-list-load-more',
+  'amp-story-page',
+  'amp-story-grid-layer',
+  'amp-story-bookend',
+];
+
+const COMPONENT_VERSIONS_PATH = path.join(project.paths.GROW_POD,
+    '/extensions/amp-component-versions.json');
+
+function getAutosuggestComponents() {
+  let result = [];
+  const components = require(COMPONENT_VERSIONS_PATH);
+  for (const component in components) {
+    if (components.hasOwnProperty(component)) {
+      result.push(component);
+    }
+  }
+  result = result.concat(BUILD_IN_COMPONENTS);
+  result = result.concat(IMPORTANT_INCLUDED_ELEMENTS);
+  return result.sort();
+}
+
+const AUTOSUGGEST_COMPONENTS = getAutosuggestComponents();
+
+examples.get('/search/autosuggest', handleAutosuggestRequest);
+examples.get('/search/do', handleSearchRequest);
+
+function handleAutosuggestRequest(request, response) {
+  const items = ({
+    items: AUTOSUGGEST_COMPONENTS,
+  });
+  response.json(items);
+}
+
+function hasExample(component) {
+  const filePath = path.join(EXAMPLE_FOLDER_PATH, component + '.html');
+  return fs.existsSync(filePath);
```

Comment:
We should avoid making a FS access. A better approach would be to write a list of samples during the build in samples builder.

Reply:
I wanted to avoid complicating the build process.
I thought since we just check for existence this is no real problem.
On my local machine it is pretty fast :-)
But of course such a sample list could more flexible and also be helpful in other places.
Since we have the `used_components` property for examples we can even have a list with examples for a component. Search could then simply pick the first one. Maybe we can fill some blanks with this logic. Although the build in components are not listed. Maybe we can simply search for them in the source code.
Let me know what you think about this.

Reply:
Just realized that we have the data already available in: `dist/examples/static/samples.json`.

Reply:
Ok, using this data now.
I changed the build so that this file is available when the dev server starts.